File: galera/tmp/PDFReader.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;

public class PDFReader {
  public static void main(String[] args) {
    String inFile = args[0] + ".pdf";
    String outFile = args[0] + ".csv";

    try {
      FileWriter fileWriter = new FileWriter(outFile, true);
      fileWriter.write("\"date\",\"code\",\"amount\",\"detail\"\n");
      PDDocument document = PDDocument.load(new File(inFile));
      int totalPages = document.getNumberOfPages();
      System.out.println("Total pages: " + totalPages);
      if (!document.isEncrypted()) {
        PDFTextStripper pdfStripper = new PDFTextStripper();
        for (int pageIndex = 1; pageIndex <= totalPages; pageIndex++) {
          pdfStripper.setStartPage(pageIndex);
          pdfStripper.setEndPage(pageIndex);
          String[] lines = pdfStripper.getText(document).split("\n");
          String date = "^\\d{2}/\\d{2}/\\d{4}$";
          String eliminate = "^Page\s+[0-9]+\s+of\s+[0-9]+";
          int column = 0;
          boolean isNew = true;
          boolean isValid = true;
          for (String line : lines) {
            if (line.matches(date)) {
              if (isNew) {
                isNew = false;
                isValid = true;
              } else {
                fileWriter.write("\"\n");
              }
              column = 1;
            } else if (line.matches(eliminate)) {
              isValid = false;
            }

            if (column == 1) {
              String[] tmp = line.split("/");
              fileWriter.write("\"" + tmp[2] + "-" + tmp[1] + "-" + tmp[0] + "\",");
              column++;
            } else if (column == 2) {
              fileWriter.write("\"" + line.replace(".", "") + "\",");
              column++;
            } else if (column == 3) {
              String tmp =
                  line.replaceFirst("\\.", "")
                      .replaceFirst("^\\s+", "")
                      .replaceFirst("\\s+", ",\"")
                      .replaceFirst("\\s+$", "");
              fileWriter.write(tmp.replace(".", ""));
              column++;
            } else if (column >= 4) {
              if (isValid) {
                fileWriter.write(" " + line.replaceFirst("^\\s+", "").replaceFirst("\\s+$", ""));
              }
            }
          }
          fileWriter.write("\"\n");
        }
      }
      document.close();
      fileWriter.close();
    } catch (IOException e) {
      e.printStackTrace();
    }
  }
}

File: galera/tmp/PDFReader.py
#!/usr/bin/env python

import csv
import os
import sys

import pdfplumber


def convert(file):
    inFile = file + ".pdf"
    outFile = file + ".csv"
    print(file, "\n", inFile, "\n", outFile)
    firstWrite = not os.path.exists(outFile)
    headers = ["date", "code", "amount", "detail"]
    with pdfplumber.open(inFile) as pdf:
        totalPages = len(pdf.pages)
        print("Total pages: ", totalPages)
        with open(outFile, "a", newline="", encoding="utf-8") as csvfile:
            writer = csv.writer(csvfile)
            if firstWrite:
                writer.writerow(headers)
            for i in range(0, totalPages, 1):
                page = pdf.pages[i]
                tables = page.extract_tables()
                for table in tables:
                    for row in table[1:]:
                        dates, codes, amounts, details = [], [], [], []
                        print(row)
                        # for j, cell in enumerate(row):
                        #     print(j, ": ", cell)
                        #     if cell is not None:
                        #         lines = cell.split("\n")
                        #         if j == 0:
                        #             dates.extend(lines[::2])
                        #             codes.extend(lines[1::2])
                        #         elif j == 2:
                        #             amounts.extend(lines)
                        #         elif j == 4:
                        #             details.extend(lines)
                        # for date, code, amount, detail in zip(
                        #     dates, codes, amounts, details
                        # ):
                        #     writer.writerow([date, code, amount, detail])


convert(sys.argv[1])
